### PR TITLE
etcd-tester: refactor checking short lived lease logic

### PR DESCRIPTION
**Pull request by fanminshi**
move the logic of waiting lease expired from stresser to checker

### CODE REVIEW

**Comment by fanminshi** [comment]
PTAL @heyitsanthony

File: tools/functional-tester/etcd-tester/checks.go [hunk 1]
```diff
@@ -87,29 +89,93 @@ func (lc *leaseChecker) Check() error {
...
+//				1. retry again by repeating steps 1 - 4 if check fails
+//				2. exit if check succeeds
+//			5. return error if all retries have failed
+//		2. return any errors on checking lease from shortLivedLeases
+func (lc *leaseChecker) shortLivedLeaseCheck() error {
+	ctx, cancel := context.WithTimeout(context.Background(), leaseCheckerTimeout)
+	errc := make(chan error)
+	defer cancel()
+	for leaseID := range lc.ls.shortLivedLeases.leases {
+		go func(id int64) {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: This is nested too deep, try:
```go
go func(id int64) { errc <- lc.checkShortLivedLease(ctx, id)  }(leaseID)
```

File: tools/functional-tester/etcd-tester/checks.go [hunk 2]
```diff
@@ -87,29 +89,93 @@ func (lc *leaseChecker) Check() error {
...
+//		1. for each lease from shortLivedLeases map
+//			1. get lease remaining TTL info.
+//			2. if lease is not expired, then wait until it expires.
+//			3. if lease is expired but not deleted, then wait for 1 second for etcd cluster to process lease deletion.
+//			4. check lease expiry.
+//				1. retry again by repeating steps 1 - 4 if check fails
+//				2. exit if check succeeds
+//			5. return error if all retries have failed
+//		2. return any errors on checking lease from shortLivedLeases
+func (lc *leaseChecker) shortLivedLeaseCheck() error {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: `func (lc *leaseChecker) checkShortLivedLeases() error` to be consistent with the other check* functions.

File: tools/functional-tester/etcd-tester/checks.go [hunk 3]
```diff
@@ -87,29 +89,93 @@ func (lc *leaseChecker) Check() error {
 		return err
 	}
 	plog.Infof("checking short lived leases %v", lc.ls.shortLivedLeases.leases)
-	return lc.check(true, lc.ls.shortLivedLeases.leases)
+	return lc.shortLivedLeaseCheck()
+
+}
+
+// shortLivedLeaseCheck verifies the expiry of shortLivedLeases.
+// the logic is follwing:
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: Why isn't `checkShortLivedLeases ensures leases expire.` enough? Listing what the function does like this is repeating what's already in there code-- are there edge cases that wouldn't be obvious from reading the code?

File: tools/functional-tester/etcd-tester/checks.go [hunk 4]
```diff
@@ -87,29 +89,93 @@ func (lc *leaseChecker) Check() error {
...
+			var err error
+			// retry in case of transient failure or lease is expired but not deleted due to the fact that etcd cluster didn't have enought time to delete it.
+			for i := 0; i < retries; i++ {
+				resp, err := lc.ls.getLeaseByID(ctx, leaseID)
+				if rpctypes.Error(err) == rpctypes.ErrLeaseNotFound {
+					break
+				}
+				if err != nil {
+					plog.Warningf("retry %d. failed to retrieve lease %v error (%v)", i, id, err)
+				} else {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: can get rid of the `else` and save some indentation,
```go
    continue
}
```

File: tools/functional-tester/etcd-tester/checks.go [hunk 5]
```diff
@@ -87,29 +89,93 @@ func (lc *leaseChecker) Check() error {
...
+func (lc *leaseChecker) shortLivedLeaseCheck() error {
+	ctx, cancel := context.WithTimeout(context.Background(), leaseCheckerTimeout)
+	errc := make(chan error)
+	defer cancel()
+	for leaseID := range lc.ls.shortLivedLeases.leases {
+		go func(id int64) {
+			var err error
+			// retry in case of transient failure or lease is expired but not deleted due to the fact that etcd cluster didn't have enought time to delete it.
+			for i := 0; i < retries; i++ {
+				resp, err := lc.ls.getLeaseByID(ctx, leaseID)
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: does this `err :=` shadow the `var err error` above?

File: tools/functional-tester/etcd-tester/checks.go [hunk 6]
```diff
@@ -87,29 +89,93 @@ func (lc *leaseChecker) Check() error {
...
+				if err != nil {
+					plog.Warningf("retry %d. failed to retrieve lease %v error (%v)", i, id, err)
+				} else {
+					if resp.TTL > 0 {
+						// sleep for the lease remaining time (represented by resp.TTL)
+						plog.Warningf("lease %v is not expired. sleep for %d until it expires.", id, resp.TTL)
+						time.Sleep(time.Duration(resp.TTL) * time.Second)
+					} else {
+						// lease is expired but not deleted might due to the fact that there aren't enough time for etcd cluster to delete expired leases.
+						// we sleep for 1 second and retry again.
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: this comment seems redundant; the two lines below already say "retry" and "Sleep(time.Second)"

File: tools/functional-tester/etcd-tester/checks.go [hunk 7]
```diff
@@ -87,29 +89,93 @@ func (lc *leaseChecker) Check() error {
...
+			for i := 0; i < retries; i++ {
+				resp, err := lc.ls.getLeaseByID(ctx, leaseID)
+				if rpctypes.Error(err) == rpctypes.ErrLeaseNotFound {
+					break
+				}
+				if err != nil {
+					plog.Warningf("retry %d. failed to retrieve lease %v error (%v)", i, id, err)
+				} else {
+					if resp.TTL > 0 {
+						// sleep for the lease remaining time (represented by resp.TTL)
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: this comment seems redundant; warning below already explains what is happening

File: tools/functional-tester/etcd-tester/checks.go (hunk 3 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/checks.go]: I was afraid that the code might not be clear. So I gave a description  to reader what the logic is before he/she start to read. I can remove the comment.

File: tools/functional-tester/etcd-tester/checks.go (hunk 2 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/checks.go]: got it.

**Comment by fanminshi** [comment]
All fixed. PTAL cc @heyitsanthony

File: tools/functional-tester/etcd-tester/checks.go [hunk 8]
```diff
@@ -87,28 +89,83 @@ func (lc *leaseChecker) Check() error {
...
-			plog.Errorf("hasLeaseExpired error: (%v)", err)
-			return err
+			plog.Warningf("retry %d. failed to retrieve lease %v error (%v)", i, leaseID, err)
+			continue
+		}
+		if resp.TTL > 0 {
+			plog.Warningf("lease %v is not expired. sleep for %d until it expires.", leaseID, resp.TTL)
+			time.Sleep(time.Duration(resp.TTL) * time.Second)
+		} else {
+			plog.Warningf("retry %d. lease %v is expired but not deleted", i, leaseID)
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/checks.go]: s/not deleted/not yet revoked/

**Comment by heyitsanthony** [comment]
please change the commit title to `etcd-tester: refactor checking short lived lease logic` and it's lgtm. Thanks!

**Pull request merged**